File: lib/core/repositories/academic/api/academic_mapper.dart
import '../academic_models.dart';
import '../academic_year_label.dart';
import 'dto/academic_class_dto.dart';
import 'dto/academic_section_dto.dart';
import 'dto/academic_teacher_assignment_dto.dart';
import 'dto/academic_year_dto.dart';

class AcademicMapper {
  const AcademicMapper();

  AcademicYear toYear(AcademicYearDto dto) {
    return AcademicYear(
      yearId: dto.yearId,
      yearLabel: normalizeAcademicYearLabel(dto.yearLabel),
      startDate: dto.startDate,
      endDate: dto.endDate,
      isCurrent: dto.isCurrent,
      status: dto.status,
    );
  }

  AcademicClass toClass(AcademicClassDto dto) {
    return AcademicClass(
      classId: dto.classId,
      academicYearId: dto.academicYearId,
      className: dto.className,
      displayOrder: dto.displayOrder,
      status: dto.status,
    );
  }

  AcademicSection toSection(AcademicSectionDto dto) {
    return AcademicSection(
      sectionId: dto.sectionId,
      classId: dto.classId,
      className: dto.className,
      sectionName: dto.sectionName,
      capacity: dto.capacity,
      strength: dto.strength,
      status: dto.status,
    );
  }

  AcademicTeacherAssignment toTeacherAssignment(
    AcademicTeacherAssignmentDto dto,
  ) {
    return AcademicTeacherAssignment(
      assignmentId: dto.assignmentId,
      teacherId: dto.teacherId,
      teacherName: dto.teacherName,
      classId: dto.classId,
      className: dto.className,
      sectionId: dto.sectionId,
      sectionName: dto.sectionName,
      role: dto.role,
      isPrimary: dto.isPrimary,
    );
  }

  List<AcademicYear> toYears(List<AcademicYearDto> dtos) =>
      [for (final dto in dtos) toYear(dto)];

  List<AcademicClass> toClasses(List<AcademicClassDto> dtos) =>
      [for (final dto in dtos) toClass(dto)];

  List<AcademicSection> toSections(List<AcademicSectionDto> dtos) =>
      [for (final dto in dtos) toSection(dto)];

  List<AcademicTeacherAssignment> toTeacherAssignments(
    List<AcademicTeacherAssignmentDto> dtos,
  ) => [for (final dto in dtos) toTeacherAssignment(dto)];
}

File: supabase/functions/_shared/academic/academic_mapper.ts
import type { AcademicYearRow } from "./academic_years_repository.ts";
import type { ClassRow } from "./classes_repository.ts";
import type { SectionListRow, SectionRow } from "./sections_repository.ts";
import type { TeacherAssignmentListRow } from "./teacher_assignments_repository.ts";

export function listEnvelope(
  items: Record<string, unknown>[],
  pagination: {
    page: number;
    pageSize: number;
    total: number;
    hasMore: boolean;
  },
): Record<string, unknown> {
  return {
    items,
    pagination: {
      page: pagination.page,
      pageSize: pagination.pageSize,
      total: pagination.total,
      hasMore: pagination.hasMore,
    },
  };
}

export function academicYearToApi(row: AcademicYearRow): Record<string, unknown> {
  return {
    yearId: row.id,
    yearLabel: row.year_label,
    startDate: row.start_date,
    endDate: row.end_date,
    isCurrent: row.is_current,
    status: row.status,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function classToApi(row: ClassRow): Record<string, unknown> {
  return {
    classId: row.id,
    academicYearId: row.academic_year_id,
    className: row.class_name,
    displayOrder: row.display_order,
    status: row.status,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function sectionToApi(
  row: SectionRow | SectionListRow,
): Record<string, unknown> {
  const className = "class_name" in row ? row.class_name : undefined;
  return {
    sectionId: row.id,
    classId: row.class_id,
    ...(className ? { className } : {}),
    sectionName: row.section_name,
    capacity: row.capacity,
    strength: row.strength,
    status: row.status,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function teacherAssignmentToApi(
  row: TeacherAssignmentListRow,
): Record<string, unknown> {
  return {
    assignmentId: row.id,
    teacherId: row.teacher_id,
    teacherName: row.teacher_name,
    classId: row.class_id,
    className: row.class_name,
    sectionId: row.section_id,
    sectionName: row.section_name,
    role: row.role,
    isPrimary: row.is_primary,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}
